src/Microsoft.AspNet.OData.Shared/TypeHelper.cs

```diff
@@ -491,5 +491,43 @@ private static Type GetInnerGenericType(Type interfaceType)
 
             return null;
         }
+
+        /// <summary>
+        /// Searches the public property with the specified name from the target type and its base types safely,
+        /// circumventing <see cref="AmbiguousMatchException"/> where a property has been redeclared in a derived type.
+        /// </summary>
+        /// <param name="targetType">The type to search from.</param>
+        /// <param name="propertyName">The name of the public property to get.</param>
+        /// <returns>An object representing the public property with the specified name from the most derived class,
+        /// if found; otherwise, null.</returns>
+        internal static PropertyInfo GetProperty(Type targetType, string propertyName)
+        {
+            PropertyInfo[] properties = targetType.GetProperties().Where(d => d.Name.Equals(propertyName)).ToArray();
```

Comment:
nitpick: Is this method used in a ho path? If so consider using using plain loops to avoid the closure lambda allocations.

Reply:
Fixed by rewriting the method for better performance and added unit tests